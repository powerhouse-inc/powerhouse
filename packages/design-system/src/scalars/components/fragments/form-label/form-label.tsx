import { Icon } from "#powerhouse";
import { cn } from "#scalars";

import {
  Tooltip,
  TooltipProvider,
} from "../../../../ui/components/tooltip/index.js";

export interface FormLabelProps
  extends React.PropsWithChildren,
    React.LabelHTMLAttributes<HTMLLabelElement> {
  required?: boolean;
  disabled?: boolean;
  description?: string;
  hasError?: boolean;
  inline?: boolean;
  className?: string;
}

export const FormLabel: React.FC<FormLabelProps> = ({
  children,
  required,
  disabled,
  description,
  hasError,
  inline,
  className,
  ...htmlLabelProps
}) => {
  const classes = cn(
    "inline-flex items-center text-sm font-medium",
    inline ? "leading-[22px]" : "leading-4",
    `text-gray-900 ${inline ? "dark:text-gray-400" : "dark:text-gray-50"}`,
    hasError && "group-hover:!text-red-900 dark:group-hover:!text-red-900",
    hasError && inline && "text-red-800 dark:text-red-800",
    hasError && !inline && "text-red-900 dark:text-red-900",
    disabled &&
      `cursor-not-allowed text-gray-700 ${
        inline ? "dark:text-gray-600" : "dark:text-gray-300"
      }`,
    inline
      ? !disabled && "group-hover:text-gray-900 dark:group-hover:text-slate-50"
      : "mb-[3px]",
    className,
  );

  if (!children) {
    // no label provided
    return null;
  }

  const extraProps = {
    ...htmlLabelProps,
  };

  return (
    <label className={classes} {...extraProps}>
      {children}
      {required && (
        <span
          className={cn(
            "ml-1 text-gray-800 group-hover:text-gray-900 dark:text-gray-400 dark:group-hover:text-slate-50",
            hasError &&
              `${inline ? "!text-red-800" : "!text-red-900"} group-hover:!text-red-900`,
          )}
        >
          *
        </span>
      )}

      {description && (
        <TooltipProvider>
          <Tooltip content={description}>
            <Icon
              name="CircleInfo"
              size={16}
              className={cn(
                "ml-1 cursor-pointer text-gray-600 hover:text-gray-500 dark:text-gray-500 dark:hover:text-gray-500",
                disabled && "text-gray-500",
              )}
            />
          </Tooltip>
        </TooltipProvider>
      )}
    </label>
  );
};
