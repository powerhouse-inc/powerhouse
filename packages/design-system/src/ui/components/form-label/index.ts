export * from "./form-label.js";
