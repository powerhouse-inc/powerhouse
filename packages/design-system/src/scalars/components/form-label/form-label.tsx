import { Icon } from "@/powerhouse/components/icon";
import React from "react";
import { twMerge } from "tailwind-merge";

export interface FormLabelProps
  extends React.PropsWithChildren,
    React.LabelHTMLAttributes<HTMLLabelElement> {
  required?: boolean;
  disabled?: boolean;
  description?: string;
  hasError?: boolean;
  className?: string;
}

export const FormLabel: React.FC<FormLabelProps> = ({
  children,
  required,
  disabled,
  description,
  hasError,
  className,
  ...htmlLabelProps
}) => {
  const classes = twMerge(
    "inline-flex items-center text-sm font-semibold text-gray-700 hover:text-gray-900 dark:text-gray-400 dark:hover:text-gray-50",
    required && "after:ml-0.5 after:text-blue-700 after:content-['*']",
    hasError &&
      "text-red-700 after:text-red-700 hover:text-red-900 hover:after:text-red-900 dark:text-red-700 dark:hover:text-red-900 dark:hover:after:text-red-900",
    disabled && "cursor-not-allowed text-gray-600",
    className,
  );

  const extraProps = {
    ...htmlLabelProps,
  };

  return (
    <label role="label" className={classes} {...extraProps}>
      {children}

      {description && (
        // TODO: add tooltip with the description
        <Icon
          name="CircleInfo"
          size={16}
          className="ml-1 cursor-pointer text-gray-600"
        />
      )}
    </label>
  );
};
