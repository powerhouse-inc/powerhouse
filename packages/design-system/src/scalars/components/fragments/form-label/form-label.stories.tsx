import type { Meta, StoryObj } from "@storybook/react";
import { FormLabel } from "./form-label.js";

const meta: Meta<typeof FormLabel> = {
  title: "Document Engineering/Fragments/FormLabel",
  component: FormLabel,
  parameters: {
    chromatic: {
      disableSnapshot: true,
    },
  },
  tags: ["autodocs"],
  argTypes: {
    required: {
      control: {
        type: "boolean",
      },
      description: "Whether the field is required",
    },
    description: {
      control: {
        type: "text",
      },
      description: "Tooltip text that appears when hovering over the info icon",
    },
    hasError: {
      control: {
        type: "boolean",
      },
      description: "Whether the label should show error styling",
    },
    disabled: {
      control: {
        type: "boolean",
      },
      description: "Whether the label is disabled",
    },
    className: {
      control: {
        type: "text",
      },
      description: "Additional CSS classes to apply to the label",
    },
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    children: "Hello, World!",
  },
};

export const WithDescription: Story = {
  args: {
    children: "Hello, World!",
    description: "This is a description",
  },
};

export const WithError: Story = {
  args: {
    children: "Hello, World!",
    hasError: true,
  },
};

export const Disabled: Story = {
  args: {
    children: "Hello, World!",
    disabled: true,
  },
};

export const Required: Story = {
  args: {
    children: "Hello, World!",
    required: true,
  },
};
