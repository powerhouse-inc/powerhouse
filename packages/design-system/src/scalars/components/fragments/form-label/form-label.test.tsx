import { fireEvent, render, screen } from "@testing-library/react";
import { FormLabel } from "./form-label.js";

describe("FormLabel Component", () => {
  it("should match snapshot", () => {
    const { asFragment } = render(
      <FormLabel data-testid="form-label">label</FormLabel>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should display the correct label text", () => {
    render(<FormLabel>Test Label</FormLabel>);
    expect(screen.getByText("Test Label")).toBeInTheDocument();
  });

  it("should forward the htmlFor attribute", () => {
    render(<FormLabel htmlFor="test-input">Label</FormLabel>);
    const label = screen.getByText("Label");
    expect(label).toHaveAttribute("for", "test-input");
  });

  it("should render children as expected", () => {
    render(
      <FormLabel>
        Label <span data-testid="child-element">Child</span>
      </FormLabel>,
    );
    expect(screen.getByText("Label")).toBeInTheDocument();
    expect(screen.getByTestId("child-element")).toBeInTheDocument();
  });

  it("should handle click events", () => {
    const handleClick = vi.fn();
    render(<FormLabel onClick={handleClick}>Clickable Label</FormLabel>);
    const label = screen.getByText("Clickable Label");
    fireEvent.click(label);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it("should display an asterisk when required", () => {
    render(<FormLabel required>Required Label</FormLabel>);
    const asterisk = screen.getByText("*");
    expect(asterisk).toBeInTheDocument();
  });

  it("should have not-allowed cursor when disabled", () => {
    render(<FormLabel disabled>Disabled Label</FormLabel>);
    const label = screen.getByText("Disabled Label");
    expect(label).toHaveClass("cursor-not-allowed");
  });

  it("should display an icon when description is provided", () => {
    render(
      <FormLabel description="This is a description">
        Label with Description
      </FormLabel>,
    );
    const icon = screen.getByTestId("icon-fallback");
    expect(icon).toBeInTheDocument();
  });

  it("should apply custom styles", () => {
    const customStyle = { color: "red", fontWeight: "bold" };
    render(
      <FormLabel data-testid="styled-label" style={customStyle}>
        Styled Label
      </FormLabel>,
    );
    const label = screen.getByTestId("styled-label");
    expect(label.style.color).toBe(customStyle.color);
    expect(label.style.fontWeight).toBe(customStyle.fontWeight);
  });
});
