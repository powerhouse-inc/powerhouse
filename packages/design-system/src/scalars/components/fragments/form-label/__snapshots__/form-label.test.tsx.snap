// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormLabel Component > should match snapshot 1`] = `
<DocumentFragment>
  <label
    class="inline-flex items-center text-sm font-medium leading-[22px] text-gray-900 group-hover:text-gray-900 dark:text-gray-400 dark:group-hover:text-slate-50 mb-[3px]"
    data-testid="form-label"
  >
    label
  </label>
</DocumentFragment>
`;
